src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -1712,15 +1710,26 @@ private final void putCharsAt(int index, CharSequence s, int off, int end) {
         }
     }
 
-    private void putStringAt(int index, String str, int off, int end) {
-        if (getCoder() != str.coder()) {
+    private void inflateIfNeededFor(String input) {
+        if (COMPACT_STRINGS && (coder != input.coder())) {
```

Comment:
I'm not completely sure whether it's a good idea in terms of maintability, but I think this can be simplified a bit more. Currently in both `String` and `ASB` we have implementation of `coder()` very much alike:
```java
// ASB
final byte getCoder() {
    return COMPACT_STRINGS ? coder : UTF16;
}

//String
byte getCoder() {
    return COMPACT_STRINGS ? coder : UTF16;
}
```
Here we have this condition
```java
if (COMPACT_STRINGS && (coder != input.getCoder())) {}
```
where the right operand of `&&` is evaluated only when `COMPACT_STRINGS` is `true` and hence it always returns the value of `coder` field. This means we can reference it directly as
```java
if (COMPACT_STRINGS && (coder != input.coder)) {}
```

Reply:
I'm not sure if this'd give us enough to motivate the refactor, especially since we'd have to widen the visibility of `String.coder`. Maybe startup could be helped a little. Either way it feels out of scope for this change, don't you think?

Reply:
Agree